File: Send-Email-SMS-Azure-AppService-MessageFlow-python/email_service.py
import json
import requests
import logging
from typing import List, Optional, Dict, Any
from dataclasses import dataclass, asdict

@dataclass
class EmailRecipient:
    email: str
    name: Optional[str] = None
    message_id: Optional[str] = None
    vars: Optional[Dict[str, Any]] = None

@dataclass
class EmailContent:
    html: Optional[str] = None
    text: Optional[str] = None
    template_id: Optional[str] = None

@dataclass
class EmailAttachment:
    file_name: str
    file_mime: str
    file_content: str
    inline: bool = False

@dataclass
class EmailRequest:
    subject: str
    smtp_account: str
    from_recipient: EmailRecipient
    to_recipients: List[EmailRecipient]
    content: EmailContent
    tags: Optional[List[str]] = None
    cc_recipients: Optional[List[EmailRecipient]] = None
    bcc_recipients: Optional[List[EmailRecipient]] = None
    reply_to: Optional[EmailRecipient] = None
    headers: Optional[Dict[str, str]] = None
    global_vars: Optional[Dict[str, Any]] = None
    attachments: Optional[List[EmailAttachment]] = None

@dataclass
class EmailResponse:
    success: bool
    message: str
    response_content: str = ""

class EmailService:
    def __init__(self, config: Dict[str, Any]):
        """
        Initialize Email service with configuration
        
        Args:
            config: Dictionary containing 'rest_api' configuration with 'authorization' and 'application_key'
        """
        self.config = config
        self.logger = logging.getLogger(__name__)
        
        # Validate configuration
        if not config.get('rest_api', {}).get('authorization'):
            raise ValueError("REST API authorization is missing in configuration")
        if not config.get('rest_api', {}).get('application_key'):
            raise ValueError("REST API application key is missing in configuration")
    
    def send_email(self, request: EmailRequest) -> EmailResponse:
        """
        Send email using MessageFlow API
        
        Args:
            request: EmailRequest object containing email details
            
        Returns:
            EmailResponse object with success status and message
        """
        self.logger.info(f"EmailService.send_email called with request: {json.dumps(asdict(request), indent=2)}")
        
        try:
            # Prepare headers
            headers = {
                'Authorization': self.config['rest_api']['authorization'],
                'Application-Key': self.config['rest_api']['application_key'],
                'Content-Type': 'application/json'
            }
            
            self.logger.info(f"Headers set - Authorization: {self.config['rest_api']['authorization'][:10]}..., Application-Key: {self.config['rest_api']['application_key'][:10]}...")
            
            # Prepare payload
            payload = {
                'subject': request.subject,
                'smtpAccount': request.smtp_account,
                'tags': request.tags,
                'content': asdict(request.content),
                'bcc': [asdict(recipient) for recipient in request.bcc_recipients] if request.bcc_recipients else None,
                'cc': [asdict(recipient) for recipient in request.cc_recipients] if request.cc_recipients else None,
                'from': asdict(request.from_recipient),
                'replyTo': asdict(request.reply_to) if request.reply_to else None,
                'headers': request.headers,
                'globalVars': request.global_vars,
                'to': [asdict(recipient) for recipient in request.to_recipients],
                'attachments': [asdict(attachment) for attachment in request.attachments] if request.attachments else None
            }
            
            # Remove None values
            payload = {k: v for k, v in payload.items() if v is not None}
            
            json_payload = json.dumps(payload, indent=2)
            self.logger.info(f"Email API Request Payload: {json_payload}")
            
            # Send request
            self.logger.info("Sending HTTP request to: https://api.messageflow.com/v2.1/email")
            
            response = requests.post(
                'https://api.messageflow.com/v2.1/email',
                headers=headers,
                json=payload,
                timeout=30
            )
            
            response_content = response.text
            self.logger.info(f"Email API Response Status: {response.status_code}")
            self.logger.info(f"Email API Response Content: {response_content}")
            
            # Parse response
            if response.status_code == 200:
                try:
                    response_data = response.json()
                    if isinstance(response_data, dict):
                        return EmailResponse(
                            success=True,
                            message=response_data.get('message', 'SMS sent successfully'),
                            response_content=response_content
                        )
                except (json.JSONDecodeError, KeyError) as e:
                    self.logger.warning(f"Cant parse response to EmailResponse model, but HTTP status is success: {e}")
                
                return EmailResponse(
                    success=True,
                    message="Email sent successfully",
                    response_content=response_content
                )
            else:
                return EmailResponse(
                    success=False,
                    message=f"HTTP Error: {response.status_code}",
                    response_content=response_content
                )
                
        except requests.exceptions.RequestException as ex:
            self.logger.error(f"HTTP request failed: {ex}")
            return EmailResponse(
                success=False,
                message=f"HTTP request failed: {str(ex)}",
                response_content=str(ex)
            )
        except Exception as ex:
            self.logger.error(f"Error sending email: {ex}")
            return EmailResponse(
                success=False,
                message=f"Error: {str(ex)}",
                response_content=str(ex)
            ) 

File: Send-Email-SMS-Azure-AppService-MessageFlow-NodeJS/MessageFlowSmsEmailApp/services/emailService.js
const axios = require('axios');

class EmailRecipient {
    constructor ({ email, name = null, messageId = null, vars = null }) {
        this.email = email;
        this.name = name;
        this.messageId = messageId;
        this.vars = vars;
    }

    toJSON() {
        const obj = { email: this.email };
        if (this.name !== null) obj.name = this.name;
        if (this.messageId !== null) obj.messageId = this.messageId;
        if (this.vars !== null) obj.vars = this.vars;
        return obj;
    }
}

class EmailContent {
    constructor ({ html = null, text = null, templateId = null }) {
        this.html = html;
        this.text = text;
        this.templateId = templateId;
    }

    toJSON() {
        const obj = {};
        if (this.html !== null) obj.html = this.html;
        if (this.text !== null) obj.text = this.text;
        if (this.templateId !== null) obj.templateId = this.templateId;
        return obj;
    }
}

class EmailAttachment {
    constructor ({ fileName, fileMime, fileContent, inline = false }) {
        this.fileName = fileName;
        this.fileMime = fileMime;
        this.fileContent = fileContent;
        this.inline = inline;
    }

    toJSON() {
        return {
            fileName: this.fileName,
            fileMime: this.fileMime,
            fileContent: this.fileContent,
            inline: this.inline
        };
    }
}

class EmailRequest {
    constructor ({
        subject,
        smtpAccount,
        from,
        to,
        content,
        tags = null,
        cc = null,
        bcc = null,
        replyTo = null,
        headers = null,
        globalVars = null,
        attachments = null
    }) {
        this.subject = subject;
        this.smtpAccount = smtpAccount;
        this.from = from;
        this.to = to;
        this.content = content;
        this.tags = tags;
        this.cc = cc;
        this.bcc = bcc;
        this.replyTo = replyTo;
        this.headers = headers;
        this.globalVars = globalVars;
        this.attachments = attachments;
    }
}

class EmailResponse {
    constructor (success, message, responseContent = '') {
        this.success = success;
        this.message = message;
        this.responseContent = responseContent;
    }
}

class EmailService {
    constructor (config) {
        if (!config?.rest_api?.authorization) {
            throw new Error('REST API authorization is missing in configuration');
        }
        if (!config?.rest_api?.application_key) {
            throw new Error('REST API application key is missing in configuration');
        }

        this.config = config;
        this.logger = console;
    }

    async sendEmail(request) {
        this.logger.info(`EmailService.sendEmail called with request: ${JSON.stringify(request, null, 2)}`);

        try {
            // Prepare headers
            const headers = {
                'Authorization': this.config.rest_api.authorization,
                'Application-Key': this.config.rest_api.application_key,
                'Content-Type': 'application/json'
            };

            this.logger.info(`Headers set - Authorization: ${this.config.rest_api.authorization.substring(0, 10)}..., Application-Key: ${this.config.rest_api.application_key.substring(0, 10)}...`);

            // Prepare payload
            const payload = {
                subject: request.subject,
                smtpAccount: request.smtpAccount,
                tags: request.tags,
                content: request.content.toJSON ? request.content.toJSON() : request.content,
                bcc: request.bcc ? request.bcc.map(r => r.toJSON ? r.toJSON() : r) : undefined,
                cc: request.cc ? request.cc.map(r => r.toJSON ? r.toJSON() : r) : undefined,
                from: request.from.toJSON ? request.from.toJSON() : request.from,
                replyTo: request.replyTo ? (request.replyTo.toJSON ? request.replyTo.toJSON() : request.replyTo) : undefined,
                headers: request.headers,
                globalVars: request.globalVars,
                to: request.to.map(r => r.toJSON ? r.toJSON() : r),
                attachments: request.attachments ? request.attachments.map(a => a.toJSON ? a.toJSON() : a) : undefined
            };

            // Remove undefined values
            Object.keys(payload).forEach(key => {
                if (payload[key] === undefined) {
                    delete payload[key];
                }
            });

            this.logger.info(`Email API Request Payload: ${JSON.stringify(payload, null, 2)}`);
            this.logger.info('Sending HTTP request to: https://api.messageflow.com/v2.1/email');

            // Send request
            const response = await axios.post(
                'https://api.messageflow.com/v2.1/email',
                payload,
                {
                    headers,
                    timeout: 30000
                }
            );

            const responseContent = JSON.stringify(response.data);

            this.logger.info(`Email API Response Status: ${response.status}`);
            this.logger.info(`Email API Response Content: ${responseContent}`);

            // Parse response
            if (response.status === 200) {
                try {
                    if (typeof response.data === 'object') {
                        return new EmailResponse(
                            true,
                            response.data.message || 'Email sent successfully',
                            responseContent
                        );
                    }
                } catch (error) {
                    this.logger.warn(`Failed to parse response to EmailResponse model, but HTTP status is success: ${error.message}`);
                }

                return new EmailResponse(
                    true,
                    'Email sent successfully',
                    responseContent
                );
            } else {
                return new EmailResponse(
                    false,
                    `HTTP Error: ${response.status}`,
                    responseContent
                );
            }

        } catch (error) {
            if (error.response) {
                // HTTP error with response
                this.logger.error(`HTTP request failed: ${error.message}`);
                return new EmailResponse(
                    false,
                    `HTTP Error: ${error.response.status}`,
                    JSON.stringify(error.response.data)
                );
            } else if (error.request) {
                // Request made but no response
                this.logger.error(`HTTP request failed: No response received - ${error.message}`);
                return new EmailResponse(
                    false,
                    `HTTP request failed: ${error.message}`,
                    error.message
                );
            } else {
                // Error setting up request
                this.logger.error(`Error sending email: ${error.message}`);
                return new EmailResponse(
                    false,
                    `Error: ${error.message}`,
                    error.message
                );
            }
        }
    }
}

module.exports = {
    EmailService,
    EmailRequest,
    EmailResponse,
    EmailRecipient,
    EmailContent,
    EmailAttachment
};
